### ISSUE TITLE
Construct RSAKeyPair from components

### ISSUE THREAD

**Issue by andrewtj** [briansmith/ring#699]
DNSSEC has an informal `key: value` text format for storing private keys which for RSA material looks like the following (from [RFC 5702](https://tools.ietf.org/html/rfc5702#section-6.1)):

```text
Private-key-format: v1.2
Algorithm:       8 (RSASHA256)
Modulus:         wVwaxrHF2CK64aYKRUibLiH30KpPuPBjel7E8ZydQW1HYWHfoGmidzC2RnhwCC293hCzw+TFR2nqn8OVSY5t2Q==
PublicExponent:  AQAB
PrivateExponent: UR44xX6zB3eaeyvTRzmskHADrPCmPWnr8dxsNwiDGHzrMKLN+i/HAam+97HxIKVWNDH2ba9Mf1SA8xu9dcHZAQ==
Prime1:          4c8IvFu1AVXGWeFLLFh5vs7fbdzdC6U82fduE6KkSWk=
Prime2:          2zZpBE8ZXVnL74QjG4zINlDfH+EOEtjJJ3RtaYDugvE=
Exponent1:       G2xAPFfK0KGxGANDVNxd1K1c9wOmmJ51mGbzKFFNMFk=
Exponent2:       GYxP1Pa7CAwtHm8SAGX594qZVofOMhgd6YFCNyeVpKE=
Coefficient:     icQdNRjlZGPmuJm2TIadubcO8X7V4y07aVhX464tx8Q=
```

I have been re-encoding these to use them with *ring* but I'd rather construct them more directly if possible. Ideally I'd like to fill in a struct provided by *ring* with the components and then pass that to a constructor on `RSAKeyPair`. Does this sound like something that'd fit in?

**Comment by briansmith** [comment]
Sure, although I've rejected doing this in the past, I am more open to doing it now.

I suggest we have a public `RSAPrivateKeyBytes` type, or similar, that looks something like this:

```rust
pub struct RSAPrivateKeyInput<'a> {
    pub n: untrusted::Input<'a>,
    pub e: untrusted::Input<'a>,
    ...
}
```

The user would fill in *all* the fields of the `RSAPrivateKeyInput` using regular record construction and then we'd implement `RSAKeyPair::from(RSAPrivateKeyInput)`. Then we'd rewrite the existing ASN.1 parsing code to be a wrapper around this new API. Then the existing tests would need to be adapted to test both the existing API and the new API. Then the new API could be exposed publicly.

I believe currently some of the sanity checking of inputs is done indirectly via the ASN.1 DER parsing code. For example, I think that the check that an input value is positive is ASN.1-specific currently. Those checks would need to be moved to `RSAKeyPair::from(RSAPrivateKeyInput)`.

**Comment by andrewtj** [comment]
In #714 it was thought better to introduce a [`der::PositiveInteger` wrapper](https://github.com/briansmith/ring/pull/714#discussion_r229506871). I'm not working on this if anyone would like to pick it up.

**Comment by briansmith** [comment]
I just added `ring::io::Positive`.

**Comment by d-e-s-o** [comment]
Let me take a stab at this issue. I need that functionality as well.

**Comment by andrewtj** [comment]
@d-e-s-o do you still need this?

**Comment by d-e-s-o** [comment]
I do. I just haven't found the time to finish up the pull request :-|

**Comment by d-e-s-o** [comment]
@briansmith have you been able to make progress on this topic? Over in [#795](https://github.com/briansmith/ring/pull/795#issuecomment-585268562) it sounded as if you were working on an implementation, but I can't seem to find it upstream. Did I miss it?

**Comment by briansmith** [comment]
@d-e-s-o Yes, I have finished the implementation. It is being tested and I hope to merge it in the next few days.

**Comment by d-e-s-o** [comment]
@briansmith has this functionality landed? I checked briefly but couldn't find it.

**Comment by briansmith** [comment]
It is in 0.17.0-alpha and I hope to ship a new release with it soon. I don't recommend using 0.17.0-alpha yet.

**Comment by briansmith** [comment]
Implemented in 0.17.0: https://docs.rs/ring/latest/ring/rsa/struct.KeyPair.html#method.from_components.

**Comment by d-e-s-o** [comment]
Thanks for getting this out. Works like a charm!
